pkg/mcs/port_set.go

```diff
@@ -0,0 +1,263 @@
+	LIGHTHOUSE_PORT_KEY_NAMED_FORMAT      = LIGHTHOUSE_PORT_KEY + ".%s"
+	LIGHTHOUSE_PORT_VALUE_FORMAT_PORT     = "%d"
+	LIGHTHOUSE_PORT_VALUE_FORMAT_PROTOCOL = LIGHTHOUSE_PORT_VALUE_FORMAT_PORT + "-%s"
+	LIGHTHOUSE_PORT_VALUE_FORMAT_FULL     = LIGHTHOUSE_PORT_VALUE_FORMAT_PROTOCOL + "-%s"
+)
+
+// PortSet encapsulates Servie ports used.
+type PortSet struct {
+	ports []mcsv1a1.ServicePort
+}
+
+// NewPortSet creates an empty PortSet.
+func NewPortSet() *PortSet {
+	return &PortSet{}
+}
+
+// NewPortSetFromServicePorts creates a PortSet populated from the ports defined
+// in v1.Service.Spec. Returns an error when the specification is invalid.
+func NewPortSetFromServicePorts(ports []corev1.ServicePort) (*PortSet, error) {
+	ps := NewPortSet()
+
+	for _, port := range ports {
+		ps.ports = append(ps.ports, mcsv1a1.ServicePort{
+			Name:        port.Name,
+			Protocol:    port.Protocol,
+			AppProtocol: port.AppProtocol,
+			Port:        port.Port,
+		})
+	}
+	return ps, nil
+}
+
+// NewPortSetFromObjectMeta creates a PortSet from the annotations defined
+// on the object's metadata. The format is parsed in accordance to the way
+// ports are encoded in Annotate(). Returns nil error and empty PortSet if
+// no port specification are found. Returns non-nil error on any parse errors.
+func NewPortSetFromObjectMeta(objmeta *metav1.ObjectMeta) (*PortSet, error) {
+	ps := NewPortSet()
+
+	for key, value := range objmeta.Annotations {
+		port, err := decodePortFromAnnotation(key, value)
+		if err != nil {
+			return nil, err
+		}
+		ps.ports = append(ps.ports, *port)
+	}
+	return ps, nil
+}
+
+// Len returns the count of ports in the PortSet.
+func (ps *PortSet) Len() int {
+	if ps == nil {
+		return 0
+	}
+	return len(ps.ports)
+}
+
+// Annotate encodes the PortSet into the object's metadata. The formatting
+// is done by encodePortAsAnnotation.
+func (ps *PortSet) Annotate(objmeta *metav1.ObjectMeta) error {
+	annotations, err := ps.AsAnnotations()
+	if err != nil {
+		return err
+	}
+
+	if objmeta.Annotations == nil {
+		objmeta.Annotations = make(map[string]string, len(annotations))
+	}
+
+	for key, value := range annotations {
+		objmeta.Annotations[key] = value
+	}
+	return nil
+}
+
+// AsAnnotations returns the needed annotations to encode the entire PortSet
+func (ps *PortSet) AsAnnotations() (map[string]string, error) {
+	annotations := make(map[string]string, len(ps.ports))
+
+	for _, p := range ps.ports {
+		key, value, err := encodePortAsAnnotation(&p)
+		if err != nil {
+			// @TODO consider if we want to encode and return a subset of
+			// the ports which were successfully encoded or fail all
+			return nil, err
+		}
+		annotations[key] = value
+	}
+
+	return annotations, nil
+}
+
+const (
+	validateOnlySharedPorts = false
+	validateAllPorts        = true
+)
+
+// Equal returns true when the PortSets are semantically identical
+func (ps *PortSet) Equals(rhs *PortSet) bool {
```

Comment:
rhs??

Reply:
C++ background ;-)
Left hand side (lhs) and Right hand side (rhs) operands names are sometimes used in operator overloading.
Think `lhs > rhs`, `lhs == rhs`, `lhs += rhs`, etc.
I had initially called it `other` but open for suggestions

Reply:
ok. I'll address that in a follow up PR where all of the Global Properties are added (Headlessness and ServiceAffinity) and use a more standard name for reading/writing from/to ObjectMeta annotations (i.e., Marshal/Unmarshal same as JSON)